programs/mango-v4/src/state/oracle.rs

```diff
@@ -317,9 +356,53 @@ pub fn oracle_state_unchecked(
                 oracle_type: OracleType::SwitchboardV1,
             }
         }
+        OracleType::OrcaCLMM => {
+            let usd_state = usdc_state_unchecked(usd_feed_opt)?;
+
+            let whirlpool = Whirlpool::try_deserialize(&mut &data[..]).unwrap();
+
+            let sqrt_price = U64F64::from_bits(whirlpool.sqrt_price);
+
+            let clmm_price = if whirlpool.token_mint_a == usdc_mint_mainnet::ID {
+                // inverted
+                let decimals = QUOTE_DECIMALS - (base_decimals as i8); // tokenMintB - tokenMintA
+                let decimal_adj = power_of_ten(decimals);
+                let inv_price = I80F48::from_num(sqrt_price * sqrt_price) * decimal_adj;
```

Comment:
On first reading this, I was worried that there might not be enough precision to do the squaring in U64F64-space in particular if we drop the last 16 bits of precision afterwards. But it looks ok: after all we assume the price in the end has good precision in I80F48.

But: did you check the decimals logic? My expectation would be that sqrt_price is already in native/native units and not in ui units, so all the decimal shifting should be unnecessary. (verify by looking at the SOL/USDC sqrt_price)

Reply:
ah yes good catch! The price is already in native/native